Made example input file which measures interface energy

jbair34 opened the pull request:
Closes issue #5867

dschwen commented:
please check your commit again

jbair34 commented:
Check it for what? it is squashed into one commit now according to my log...

permcody commented:
It still contains a libMesh update. We can't merge that. You'll need to soft reset your commit and remove your libMesh update.

jbair34 commented:
@permcody So I should git reset HEAD^ and then re-commit it and push it?

permcody commented:
Almost...

After you reset HEAD^, you'll want to run `git submodule update`. This will checkout the existing submodule hash. You can verify by doing another `git status` where you should only see your other file sitting there which you can then recommit and push.

Commit:
reset then updated libmesh submodule and recommiting example #5867

jbair34 commented:
Thanks @permcody  I think that commit should be better... I didn't see anything out of place in the git status...

dschwen commented:
It would be nice if we could figure out a way to make the size of the cell easily configurable. Right now it appears in 
* the initial condition
* the Int_energy MOOSE function 
* one PointValue postprocessor

One suggestion would be to use ```SideAverageValue``` instead of ```Pointvalue```. this would permit you to specify boundary names, instead of the domain size dependent point.

dschwen commented:
Ok, to recap: Use an ```ElementIntegralMaterialProperty``` with ```mat_prop = 1``` to calculate the length of your simulation cell and write a ramp IC that creates a linear ramp from a min to max value across the simulation cell (query the x extent of the mesh for this).

jbair34 commented:
So I am working on the IC but quick question... ElementIntegralMaterialProperty is again a post processor so won't it not work just like the numelems post processor didn't work?

dschwen commented:
But we will not be using that PP in the IC! It is for use in the function that computes the interfacial free energy.

jbair34 commented:
Right... got it.

jbair34 commented:
Is there an example somewhere for how to query the x extent of the mesh? I haven't found anything in the other ICs like that...

dschwen commented:
```
MooseMesh::dimensionWidth(unsigned int component)
```
is what you need. Search through the code to find out how to access the mesh object.

Commit:
Created RampIC and simplified Interface example so that only mesh needs to be changed #5867

Switched Header on RampIC #5867

Fixed RampIC to work for any xmin xmax mesh

Changed name of post processor from xmax to volume

permcody commented:
@jbair34 - you should probably squash these down again. At least the commits that represent bug fixes to your PR.

dschwen commented on modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i:
```diff
@@ -0,0 +1,194 @@
+[]
+
+[Postprocessors]
+  # The total free enrgy of the simulation cell to observe the energy reduction.
```
typo: enrgy

dschwen commented on modules/phase_field/include/ics/RampIC.h:
```diff
@@ -0,0 +1,56 @@
+  Real _xlength;
+  Real _xmin;
+  Real _value_left;
+  Real _value_right;
```
those members can all be ```const```

dschwen commented on modules/phase_field/examples/measure_interface_energy/1Dinterface_energy.i:
```diff
@@ -0,0 +1,194 @@
+    block = 0
+    mat_prop = 1
+  [../]
+  # Find concentration in each phase using max and min
```
No, this is _wrong_! We do not know a priori if the concnetrations are the min and max, or the max and min respectively! You need to be consistent with the Fleft and Fright PPs! Use SideAverageValue.

jwpeterson replied:
> No, this is wrong!

I think you mean "Nein!"

dschwen replied:
9
9
9
9
!!

dschwen commented:
Please add a test for the RampIC as well (preferably with an odd simulation cell xmin = -5 xmax = 13)

Commit:
Added test for RampIC and fixed some bugs

fixed typo in 1Dinterface example problem

Made RampIC test to solve=false and switched const real around

Removed RampIC_test kernels

Fixed several comments and aesthetics

Missed one

dschwen commented:
We test ICs in the phase field module with ```solve = false```. Take a look at some of the tests there. There is no need to have any kernels and any solve in these tests.

jbair34 commented:
I updated that IC but just so you know in phase_field/tests/CH_IC there are a whole bunch of IC tests that are Transient with a couple of time steps running... Do you want me to change all of those to solve = false?

dschwen commented:
No, we'll do a separate issue for those.

dschwen commented:
I think @permcody fixed the option to disable the kernel coverage check. If so you can remove all kernels (add ```kernel_coverage_check = false``` in the ```[Problem]``` block). You can remove the ```w``` variable and its kernels anyways.

permcody commented:
@dschwen - another review please!

dschwen commented on modules/phase_field/tests/RampIC/RampIC_test.i:
```diff
@@ -0,0 +1,57 @@
+    derivative_order = 2
+    block = 0
+  [../]
+[]
```
I don't think we need a Materials block here.

dschwen commented on modules/phase_field/src/base/PhaseFieldApp.C:
```diff
@@ -293,6 +294,7 @@ PhaseFieldApp::registerObjects(Factory & factory)
   registerInitialCondition(SpecifiedSmoothCircleIC);
   registerInitialCondition(ThumbIC);
   registerInitialCondition(Tricrystal2CircleGrainsIC);
+  registerInitialCondition(RampIC);
```
The include and register statements are sorted alphabetically. Let's keep it that way.

dschwen commented on modules/phase_field/include/ics/RampIC.h:
```diff
@@ -0,0 +1,56 @@
+  /**
+   * Evaluate the function at the current quadrature point and timestep.
+   */
+  const Real f();
```
What is this? It does not appear in the implementation file.

dschwen commented on modules/phase_field/include/ics/RampIC.h:
```diff
@@ -0,0 +1,56 @@
+InputParameters validParams<RampIC>();
+
+/**
+ * Defines a boundary condition that forces the value to be a user specified
```
No, it does not. Please make sure that copy pasted comments are updated to reflect what the code actually does.

dschwen commented on modules/phase_field/include/ics/RampIC.h:
```diff
@@ -0,0 +1,56 @@
+#include "InitialCondition.h"
+#include "InputParameters.h"
+#include "FEProblem.h"
+#include "MooseMesh.h"
```
neither MooseMesh.h nor FEProblem.h need to be included in the header. We should avoid unnecessary include statements as they will increase compile time.

dschwen commented on modules/phase_field/tests/RampIC/RampIC_test.i:
```diff
@@ -0,0 +1,40 @@
+  type = Steady
+  solve_type = PJFNK
+  petsc_options_iname = '-pc_type -pc_hypre_type'
+  petsc_options_value = 'hypre boomeramg'
```
if we do not solve we won't need PETSc options...

dschwen commented on modules/phase_field/tests/RampIC/tests:
```diff
@@ -0,0 +1,8 @@
+[Tests]
+  [./RampIC_test]
```
I'm not a fan of tests that are called ```*test```. But let's just leave it that way. I plan on reorganizing the entire phase_field module tests directory anyways... if I end up with too much time on my hands.
